protocols/gossipsub/src/behaviour.rs

```diff
@@ -1623,15 +1628,22 @@ where
         }
 
         for p in px {
-            // TODO: Once signed records are spec'd: extract signed peer record if given and handle
-            // it, see https://github.com/libp2p/specs/pull/217
-            if let Some(peer_id) = p.peer_id {
+            let dial_opts = if let Some(signed_peer_record) = p.signed_peer_record {
+                // If signed envelop is invalid, return None and ignore peer exchange
+                PeerRecord::from_signed_envelope(signed_peer_record)
+                    .ok()
+                    .map(|peer_record| (peer_record.peer_id(), peer_record.addresses().to_vec()))
+            } else {
+                p.peer_id.map(|peer_id| (peer_id, vec![]))
+            };
+
+            if let Some((peer_id, addresses)) = dial_opts {
                 // mark as px peer
                 self.px_peers.insert(peer_id);
 
                 // dial peer
                 self.events.push_back(NetworkBehaviourAction::Dial {
-                    opts: DialOpts::peer_id(peer_id).build(),
+                    opts: DialOpts::peer_id(peer_id).addresses(addresses).build(),
```

Comment:
I know this should have been done earlier, probably on line 1617. 

I think we probably don't want to try to connect to all the peers, we probably only want to try and get up to `mesh_n_high`. I'd suggest we set `n` on line 1617 to be the min of `self.config.prune_peers()` or `self.mesh.get(&topic_hash).map(|x| x.len()).unwrap_or(0)` or something. Which may mean calculating this in `handle_prune()` because we don't have the topic hash here.

Reply:
Agree with the needed change in general. 

Note that `prune_peers` according to our docs configure the number of peers we _send_ to others, not the number of peers we connect to. Either code or docs need to be adjusted. But using the current docs, this should probably not be used here (or in `handle_prune`)

It's also not clear to me if handling the prune is the best place to connect to those peers. We already graft peers in the heartbeat when below `mesh_n_low`. We might want to connect to these peers to accelerate the process, or we might want to prioritize already connected peers since they are somewhat more trusted. In any case, both `mesh_n_high` and `mesh_n_low` should probably be used to select the number of peers from this set we want to try.

Reply:
PeerExchange is meant to give you more peers in case you are only connected to a few, right?

Perhaps it would be better to just add those to some kind of list and let the grafting logic upon heartbeats do the connecting?

For the case where we don't have many peers, that would otherwise "fail" in that we can't connect to as many peers as we'd like.

Reply:
Those are good points, and path forward is opinionated. Maintainers and @divagant-martian @AgeManning could you comment on what makes more sense?

IMO connecting eagerly instead of accumulating keep it simpler + limit only up to mesh needs. For reference Go impl eagerly schedules all peer exchanges to the connect channel (with max cap MaxPendingConnections)

Reply:
I've not touched gossipsub much so I'll leave it to @mxinden and @AgeManning to make a decision here.

> For reference Go impl eagerly schedules all peer exchanges to the connect channel (with max cap MaxPendingConnections)

That is an interesting data point.

Reply:
Yeah, now that I think about this, there's not that many behaviours that dial peers. 

In rust-libp2p, when we dial a peer, we not only connect to this behaviour, but we connect to all behaviours that the peer supports. In Lighthouse, we are pretty strict about which peers we dial. We have limits on outbound vs inbound connections, banned ips, scoring etc and so have very stringent dialing conditions. It would actually kind of cause havoc for us if gossipsub was out and about dialing random peers that are just feed to us. In principle perhaps they could form eclipse attacks but feeding us malicious peers constantly. 

I think there are two ways forward and I'd be curious to hear everyone's opinion on them, because I dont have a strong preference:
1) We keep it as is an dial eagerly provided we have the need, i.e that `mesh_high` > `connected_peers[topic_hash].len()` (we could potentially use `mesh_n` but probably high is better). This is somewhat safe, because px is disabled by default in the config and turning it on should be done at one's own risk. 
2) We create a new event type. Gossipsub can propagate upwards potential peers to the swarm and the user can decide whether to dial or not. 

After writing these out, I think 1 is probably the easiest as I realize its off by default and can be opted in to.

Reply:
(2) is also touched on in https://github.com/libp2p/rust-libp2p/issues/2680. At the moment, there is no way for a behaviour to in a standard way emit an event that it learned addresses about a peer. This makes it difficult to build a general address-book style behaviour.

I do think it would be a good idea but the details would need to be fleshed out. One idea would be to immediately re-inject this information into the behaviour, therefore effectively distributing it across all behaviours. One of them could then be a (persistent) address book that uses the provided addresses to enhance future dial attempts.

Reply:
For the sake of unblocking this PR we can go with eager connection, since as Age mentioned, this needs to be explicitly turned on.

Reply:
@AgeManning Should `do_px` be used to not consume px by default? Or add a new config param like `use_px` to split provision and consumption of px

Reply:
I think we should have a more explicit config. I think we should opt for `enable_peer_exchange()` which enables both sending and receiving. I dont think users really need granularity over one or the other (we can add it later if its requested). 

I'd suggest renaming `do_px` to `enable_peer_exchange`. Have it off my default.

Reply:
@dapplion to achieve this you technically need to change nothing. This is already the case. Age's suggestion is a nice to have, but would change the public api, so the version should be updated accordingly. I'm fine with doing or not doing this change

Reply:
We can do it in a non-breaking way by adding `enable_peer_exchange` on `Config` and deprecating the `do_px` function. Internally, the field can be renamed without issues.